keps/sig-apps/4212-declarative-node-maintenance/README.md

```diff
@@ -0,0 +1,2168 @@
+- Mirror pods are skipped.
+- Terminating pods are skipped.
+- Pods and ReplicaSets/ReplicationControllers without owning controllers are blocking by default
+  (the check can be modified with the `skip-nodes-with-custom-controller-pods` option).
+- System pods (in the `kube-system` namespace) without a matching PDB are blocking by default
+  (the check can be modified with the `skip-nodes-with-system-pods` option).
+- Pods with `cluster-autoscaler.kubernetes.io/safe-to-evict: "false"` annotation are blocking.
+- Pods with local storage are blocking unless they have a
+  `cluster-autoscaler.kubernetes.io/safe-to-evict-local-volumes` annotation
+  (the check can be modified with the `skip-nodes-with-local-storage` option, e.g., this check
+  is skipped on [AKS](https://learn.microsoft.com/en-us/azure/aks/cluster-autoscaler?tabs=azure-cli#cluster-autoscaler-profile-settings)).
+- Pods with PDBs that do not have `disruptionsAllowed` are blocking.
+
+This can be enhanced with other rules and overrides.
+
+It uses this logic to first check if all pods can be removed from a node. If not, it will report
+those nodes. Then it will group all pods by a priority and evict them gradually from the lowest
+to highest priority. This may include DaemonSet pods.
+
+### kubelet
+
+Graceful Node Shutdown is a part of the current solution for node maintenance. Unfortunately, it is
+not possible to rely solely on this feature as a go-to solution for graceful node and workload
+termination.
+
+- The Graceful Node Shutdown feature is not application aware and may prematurely disrupt workloads 
+  and lead to data loss.
+- The kubelet controls the shutdown process using Dbus and systemd, and can delay (but not entirely
+  block) it using the systemd inhibitor. However, if Dbus or the kubelet is restarted during the
+  node shutdown, the shutdown might not be registered again, and pods might be terminated
+  ungracefully. Also, new workloads can get scheduled on the node while the node is shutting down.
+  Cluster admins should, therefore, plan the maintenance in advance and ensure that pods are
+  gracefully removed before attempting to shut down or restart the machine.
+- The kubelet has no way of reliably detecting ongoing maintenance if the node is restarted in the
+  meantime.
+- Graceful termination of static pods during a shutdown is not possible today. It is also not
+  currently possible to prevent them from starting back up immediately after the machine has been
+  restarted and the kubelet has started again, if the node is still under maintenance. 
+
+### Motivation Summary
+
+To sum up. Some applications solve the disruption problem by introducing validating admission
+webhooks. This has some drawbacks. The webhooks are not easily discoverable by cluster admins. And
```

Comment:
Is this not solved with the new addition of CEL?

Reply:
CEL cannot do everything. For example, it cannot check additional resources, call other APIs. Also the eviction API was not designed to support this, so it is still an unexpected behavior for the clients.

Reply:
I see, I fully agree webhooks are something we should move away